### PR TITLE
goas: switch balance commitments to hash based commitments

### PR DESCRIPTION
**Pull request by davidrusu**

### CODE REVIEW

File: goas/cl/cl/src/balance.rs [hunk 1]
```diff
@@ -1,164 +1,156 @@
...
-        // Zero is the same across all units
-        let (nmo, eth) = (unit_point("NMO"), unit_point("ETH"));
-
-        let mut rng = rand::thread_rng();
-        let b = BalanceWitness::random(&mut rng);
-        assert_eq!(
-            b.commit([&NoteWitness::basic(0, nmo)], []),
-            b.commit([&NoteWitness::basic(0, eth)], []),
-        );
+    pub fn insert_positive(&mut self, unit: Unit, value: Value) {
```
Comment (zeegomo) [inline goas/cl/cl/src/balance.rs:69]: would it be much slower to use a map?

File: goas/cl/cl/src/balance.rs [hunk 2]
```diff
@@ -1,164 +1,156 @@
...
 
-        let note = NoteWitness::basic(10, nmo);
+    pub fn combine(balances: impl IntoIterator<Item = Self>, blinding: [u8; 16]) -> Self {
+        let mut combined = BalanceWitness::zero(blinding);
 
-        let a = bal_a.commit([&note], []);
-        let b = bal_b.commit([&note], []);
+        for balance in balances {
+            for unit_bal in balance.balances.iter() {
+                if unit_bal.pos > unit_bal.neg {
```
Comment (zeegomo) [inline goas/cl/cl/src/balance.rs:125]: is it an invariant that only one of the negative or positive value is set?

File: goas/cl/cl/src/balance.rs (hunk 1 above)
Comment (davidrusu) [reply goas/cl/cl/src/balance.rs:69]: idk how much slower, but we don't expect many units in a ptx, and O(n) scanning a list is usually faster than binary search/hashing/btree for small n.

File: goas/cl/cl/src/balance.rs (hunk 2 above)
Comment (davidrusu) [reply goas/cl/cl/src/balance.rs:125]: no, it's not, but we only have helper functions for inserting a negative or positive u64.

File: goas/cl/cl/src/balance.rs [hunk 3]
```diff
@@ -1,164 +1,156 @@
...
-        let b = BalanceWitness::random(&mut rng);
-        assert_eq!(
-            b.commit([&NoteWitness::basic(0, nmo)], []),
-            b.commit([&NoteWitness::basic(0, eth)], []),
-        );
+    pub fn insert_positive(&mut self, unit: Unit, value: Value) {
+        let mut found = false;
+        for unit_bal in self.balances.iter_mut() {
+            if unit_bal.unit == unit {
+                found = true;
```
Comment (zeegomo) [inline goas/cl/cl/src/balance.rs:73]: can we return instead of break + flag? It's still not constant time

File: goas/cl/cl/src/balance.rs (hunk 2 above)
Comment (zeegomo) [reply goas/cl/cl/src/balance.rs:125]: yeah sorry misunderstood

File: goas/cl/cl/src/balance.rs (hunk 1 above)
Comment (zeegomo) [reply goas/cl/cl/src/balance.rs:69]: mainly to avoid coding this by hand but it's quite easy so no problem

File: goas/cl/cl/src/balance.rs (hunk 3 above)
Comment (davidrusu) [reply goas/cl/cl/src/balance.rs]: yeah, that's better

File: goas/cl/cl/src/balance.rs (hunk 1 above)
Comment (davidrusu) [reply goas/cl/cl/src/balance.rs:69]: yeah, it's not too bad.

**Review by zeegomo — APPROVED** [review]
This is still variable time and might leak things, but it's still an improvement and we can solve this later

**Pull request merged**